metagpt/actions/write_code.py

```diff
@@ -135,20 +156,44 @@ async def run(self, *args, **kwargs) -> CodingContext:
         return coding_context
 
     @staticmethod
-    async def get_codes(task_doc, exclude) -> str:
+    async def get_codes(task_doc, exclude, mode="normal") -> str:
         if not task_doc:
             return ""
         if not task_doc.content:
             task_doc.content = FileRepository.get_file(filename=task_doc.filename, relative_path=TASK_FILE_REPO)
         m = json.loads(task_doc.content)
-        code_filenames = m.get("Task list", [])
+        code_filenames = m.get("Task list", []) if mode == "normal" else m.get("Refined Task list", [])
         codes = []
         src_file_repo = CONFIG.git_repo.new_file_repository(relative_path=CONFIG.src_workspace)
-        for filename in code_filenames:
-            if filename == exclude:
-                continue
-            doc = await src_file_repo.get(filename=filename)
-            if not doc:
-                continue
-            codes.append(f"----- {filename}\n" + doc.content)
+
+        if mode == "guide":
+            src_files = src_file_repo.all_files
+            old_file_repo = CONFIG.git_repo.new_file_repository(relative_path=CONFIG.old_workspace)
+            old_files = old_file_repo.all_files
+            union_files_list = list(set(src_files) | set(old_files))
+            for filename in union_files_list:
+                if filename == exclude:
+                    if filename in old_files and filename != "main.py":
```

Comment:
why exclude `main.py` ?

Reply:
In most cases, main.py does not contain too much code, and it is not necessary to refer to legacy main.py to generate good code. The existence of legacy main.py may lead to the generation of a lot of unnecessary content.

Reply:
Usually, the llm will generate code in one file for small projects. It's our requirement that split the `main.py` and `other_sourcecode.py`. May here should add an explanation about excluding.